File: Python/160-Intersection-of-Two-Linked-Lists.py
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        nodeA, nodeB = headA, headB
        cache = {}

        while nodeA or nodeB:
            if nodeA:
                if nodeA in cache:
                    return nodeA
                else:
                    cache[nodeA] = nodeA
                    nodeA = nodeA.next
            if nodeB:
                if nodeB in cache:
                    return nodeB
                else:
                    cache[nodeB] = nodeB
                    nodeB = nodeB.next
        return None

File: JavaScript/160-intersection-of-two-linked-lists.js
let getIntersectionNode = (headA, headB) => {
  let currA = headA
  let currB = headB
  let aReached = (bReached = false)
  while (currA && currB) {
    if (currA === currB) {
      return currA
    }
    currA = currA.next
    currB = currB.next
    if (!currA && !aReached) {
      aReached = true
      currA = headB
    }
    if (!currB && !bReached) {
      bReached = true
      currB = headA
    }
  }
  return null
}
